packages/codemods/src/plugins/types.ts

```diff
@@ -0,0 +1,95 @@
+import { ImportDeclaration, JSCodeshift, ASTNode } from 'jscodeshift';
+
+// We want to ensure the config contract is correct so devs can get type safety
+type ValidateConfig<T, Struct> = T extends Struct
+  ? Exclude<keyof T, keyof Struct> extends never
+    ? T
+    : never
+  : never;
+
+export interface PluginMetadata {
+  name: string;
+}
+
+type BaseConfig = { processedPlugins: Array<PluginMetadata>; j: JSCodeshift };
+
+/**
+ * Interface for codemods that handle migration from CSS-in-JS libraries to Compiled
+ */
+export interface MigrationTransformer {
+  /**
+   * Build the compiled import replacing the existing import
+   *
+   * @param config The configuration object
+   * @param config.j The JSCodeshift object
+   * @param config.currentNode The existing import node that will be replaced
+   * @param config.defaultSpecifierName The import name
+   * @param config.namedImport The export from Compiled to be imported
+   * @param config.compiledImportPath The import path for Compiled
+   *
+   * @returns The import to replace config.currentNode
+   */
+  buildImport?<T>(
+    config: ValidateConfig<
+      T,
+      BaseConfig & {
+        originalNode: ImportDeclaration;
+        currentNode: ImportDeclaration;
+        defaultSpecifierName: string;
+        namedImport: string;
+        compiledImportPath: string;
+      }
+    >
+  ): ImportDeclaration;
+
+  /**
+   * Insert AST nodes before the compiled import
+   *
+   * @param config The configuration object
+   * @param config.j The JSCodeshift object
+   * @param config.newImport The new import node replaced in `buildImport`
+   *
+   * @returns Nodes to insert before the import
+   */
+  insertBeforeImport?<T>(
```

Comment:
I still think `insertBeforeImport` and `insertAfterImport` could just be replaced with having access to the program. The babel plugin in compiled uses `visitor: { Program: { exit(path, state) {} } }` to do final insertions, including inserting at particular imports.

Reply:
Done - just added `program` as the only entrypoint for now